addons/mail/models/mail_render_mixin.py

```diff
@@ -413,31 +384,30 @@ def _render_template_jinja(self, template_txt, model, res_ids,
         if not template_txt:
             return results
 
-        # try to load the template
-        try:
-            jinja_env = jinja_safe_template_env if no_autoescape else jinja_template_env
-            template = jinja_env.from_string(tools.ustr(template_txt))
-        except Exception:
-            _logger.info("Failed to load template %r", template_txt, exc_info=True)
-            return results
-
         # prepare template variables
-        variables = self._render_jinja_eval_context()
+        variables = self._render_qweb_eval_context()
         if add_context:
             variables.update(**add_context)
-        safe_eval.check_values(variables)
-
 
         for record in self.env[model].browse(res_ids):
             variables['object'] = record
-            try:
-                render_result = template.render(variables)
-            except Exception as e:
-                _logger.info("Failed to render template : %s", e, exc_info=True)
-                raise UserError(_("Failed to render template : %s", e))
-            if render_result == u"False":
-                render_result = u""
-            results[record.id] = Markup(render_result)
+
+            results[record.id] = ''
+
+            instructions = parse_small_qweb(tools.ustr(template_txt))
+            instructions
+            for string, expression in instructions:
+                results[record.id] += string
+
+                if expression:
+                    try:
+                        result = safe_eval.safe_eval(expression, variables)
```

Comment:
> Assuming expression is an expression, why use safe_eval (which allows statements) rather than expr_eval (which only allows expressions)?

expr_eval does not allow to provide a dictionary a variables.

> And why even allow arbitrary expressions?

Because we want to keep the behavior that interpolate `Hello ${expr}` in email template for some fields (subect, email_to, ...).
see: https://github.com/odoo/odoo/pull/70045#issuecomment-866107167